File: packages/shared-types/src/schemas/patient.ts
import { z } from "zod";
import { UuidSchema } from "./common.js";

export const MaritalStatusSchema = z.enum([
  "single",
  "married",
  "divorced",
  "widowed",
  "separated",
  "domestic_partner",
  "other",
]);

export const PatientSchema = z.object({
  id: UuidSchema,
  practiceId: UuidSchema,
  firstName: z.string().min(1).max(100),
  lastName: z.string().min(1).max(100),
  dateOfBirth: z.string().date(), // ISO 8601 date string, YYYY-MM-DD
  sex: z.enum(["male", "female", "other", "unknown"]).optional(),
  maritalStatus: MaritalStatusSchema.optional(),
  lastXrayDate: z.string().date().optional(),
  phone: z.string().max(20).optional(),
  email: z.string().email().max(255).optional(),
  addressLine1: z.string().max(255).optional(),
  addressLine2: z.string().max(255).optional(),
  city: z.string().max(100).optional(),
  state: z.string().length(2).optional(),
  zip: z.string().max(10).optional(),
  // PHI — SSN stored AES-256 encrypted at application layer, never plaintext.
  // Accepts 4-digit (last four) or full 9-digit SSN.
  ssn: z.string().regex(/^\d{4}$|^\d{9}$/).optional(),
  emergencyContactName: z.string().max(200).optional(),
  emergencyContactPhone: z.string().max(20).optional(),
  occupation: z.string().max(200).optional(),
  employer: z.string().max(200).optional(),
  referralSource: z.string().max(200).optional(),
  allergies: z.array(z.string()).default([]),
  medicalAlerts: z.array(z.string()).default([]),
  medications: z.array(z.string()).default([]),
  doctorNotes: z.string().max(5000).optional(),
  smsOptOut: z.boolean().default(false),
  // Soft delete
  deletedAt: z.string().datetime().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const CreatePatientSchema = PatientSchema.omit({
  id: true,
  deletedAt: true,
  createdAt: true,
  updatedAt: true,
}).extend({
  // practiceId is inferred from the authenticated X-Practice-ID header;
  // providing it in the body is optional and used only for mismatch validation.
  practiceId: UuidSchema.optional(),
});

export const UpdatePatientSchema = CreatePatientSchema.partial();

export const PatientSearchQuerySchema = z.object({
  q: z.string().min(1).optional(),
  page: z.number().int().min(1).default(1),
  pageSize: z.number().int().min(1).max(100).default(20),
});

export type MaritalStatus = z.infer<typeof MaritalStatusSchema>;
export type Patient = z.infer<typeof PatientSchema>;
export type CreatePatient = z.infer<typeof CreatePatientSchema>;
export type UpdatePatient = z.infer<typeof UpdatePatientSchema>;
export type PatientSearchQuery = z.infer<typeof PatientSearchQuerySchema>;

File: apps/api/app/models/patient.py
from __future__ import annotations

import uuid
from datetime import date

from sqlalchemy import Boolean, CheckConstraint, Date, Index, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, BYTEA, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, PHIMixin


class Patient(Base, PHIMixin):
    __tablename__ = "patients"

    practice_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        # No FK here — practices lives in the same DB but we avoid a hard FK
        # on PHI tables to make cross-schema moves easier later.
        nullable=False,
        index=True,
    )

    # Demographics
    first_name: Mapped[str] = mapped_column(String(100), nullable=False)
    last_name: Mapped[str] = mapped_column(String(100), nullable=False)
    date_of_birth: Mapped[date] = mapped_column(Date, nullable=False)
    sex: Mapped[str | None] = mapped_column(String(10), nullable=True)
    marital_status: Mapped[str | None] = mapped_column(String(20), nullable=True)
    last_xray_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    last_dental_visit: Mapped[str | None] = mapped_column(Text, nullable=True)
    previous_dentist: Mapped[str | None] = mapped_column(String(200), nullable=True)

    # Emergency contact
    emergency_contact_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    emergency_contact_phone: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # Employment / referral
    occupation: Mapped[str | None] = mapped_column(String(200), nullable=True)
    employer: Mapped[str | None] = mapped_column(String(200), nullable=True)
    referral_source: Mapped[str | None] = mapped_column(String(200), nullable=True)

    # Contact
    phone: Mapped[str | None] = mapped_column(String(20), nullable=True)
    email: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Address
    address_line1: Mapped[str | None] = mapped_column(Text, nullable=True)
    address_line2: Mapped[str | None] = mapped_column(Text, nullable=True)
    city: Mapped[str | None] = mapped_column(String(100), nullable=True)
    state: Mapped[str | None] = mapped_column(String(2), nullable=True)
    zip: Mapped[str | None] = mapped_column(String(10), nullable=True)

    # PHI — AES-256-GCM encrypted at the application layer before storage.
    # Accepts last-4 digits or full 9-digit SSN.
    ssn_encrypted: Mapped[bytes | None] = mapped_column(BYTEA, nullable=True)

    # Clinical flags
    allergies: Mapped[list[str]] = mapped_column(
        ARRAY(Text),
        nullable=False,
        server_default="{}",
    )
    medical_alerts: Mapped[list[str]] = mapped_column(
        ARRAY(Text),
        nullable=False,
        server_default="{}",
    )
    medications: Mapped[list[str]] = mapped_column(
        ARRAY(Text),
        nullable=False,
        server_default="{}",
    )
    dental_symptoms: Mapped[list[str]] = mapped_column(
        ARRAY(Text),
        nullable=False,
        server_default="{}",
    )
    doctor_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    sms_opt_out: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        server_default="false",
    )

    __table_args__ = (
        CheckConstraint(
            "sex IN ('male', 'female', 'other', 'unknown')",
            name="ck_patients_sex",
        ),
        CheckConstraint(
            "marital_status IN ('single','married','divorced','widowed',"
            "'separated','domestic_partner','other')",
            name="ck_patients_marital_status",
        ),
        # Search: last_name-first_name prefix search within a practice
        Index("ix_patients_practice_name", "practice_id", "last_name", "first_name"),
        # List with soft-delete filter
        Index("ix_patients_practice_deleted", "practice_id", "deleted_at"),
    )
